翻译glTF模型和图像位置解析算法 (#1790)
cocos-creator/creator-docs: en/asset/model/glTF.md

# glTF Models

Cocos Creator supports glTF 2.0 and earlier file formats.

## URI parsing

> Creator supports URIs of the following form specified in glTF:

Reviewer: Creator supports URIs in the following form specified in glTF

> Creator supports URIs of the following form specified in glTF:

Author: ```suggestion
Creator supports URIs in the following form specified in glTF:

```

The prefab will use a node without any spatial transformation information as the root node, and all [root nodes](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#scenenodes) of the glTF scene will be the children of this node.

#### Node Conversion

The properties in the glTF node will be converted to properties in the prefab node according to the texture relationships in the following table:

| glTF Node Property | Prefab Node Property |
| :----------- | :----------- |
| [Hierarchy](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodechildren) | Hierarchy |
| [Displacement](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodetranslation) | Position |
| [Rotation](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#noderotation) | Rotation |
| [Scaling](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodescale) | Scaling |
| [Matrix](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodematrix) | Decompress and set the position, rotation, and scaling respectively |
| [Mesh](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodemesh) | MeshRenderer component |

> | [Skin](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodeskin) | Skinned Mess Renderer component |

Reviewer 2: ```suggestion
| [Skin](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodeskin) | SkinnedMessRenderer component |

```

| [Initial Weight](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodeweights) | (Skinned) MeshRenderer component weight |

#### MeshRenderer

If the glTF node references a mesh, then the corresponding prefab node will also have a MeshRenderer component added to it after import. If the glTF node also references a skin, then the corresponding prefab node will also have a SkinnedMeshRenderer added to it.

The mesh, skeletons, and materials in the SkinnedMeshRenderer component correspond to the transformed glTF mesh, skin, and material assets.

> If the glTF node specifies an initial weight, the converted (skinned) mesh renderer will also carry this weight.

Reviewer 2: ```suggestion
If the glTF node specifies an initial weight, the converted (skinned) meshrenderer will also carry this weight.

```

# glTF Models

Cocos Creator supports glTF 2.0 and earlier file formats.

## URI parsing

Creator supports URIs in the following form specified in glTF:

- Data URI

- Relative URI path

- File URL

- File path

## Conversion Relationships

When importing a glTF model into Creator, the assets in glTF will be converted to assets in Creator according to the following relationships:

> | glTF assets | Cocos Creator assets |

Reviewer 2: 我记得表格这里是要用单数的？

> | glTF assets | Cocos Creator assets |

Author: ```suggestion
| glTF Asset | Cocos Creator Asset |

```

| [glTF Material](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-material) | Material |
| [glTF Texture](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-texture) | Texture |
| [glTF Image](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-image) | Image |
| [glTF Animation](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#reference-animation) | Animation Clip |

### glTF Scene

After import, the glTF scene will be converted to a prefab asset in Creator, and the nodes recursively contained in the glTF scene will be converted to nodes in the prefab one by one according to the same hierarchical relationship.

#### Scene Root Node

The prefab will use a node without any spatial transformation information as the root node, and all [root nodes](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#scenenodes) of the glTF scene will be the children of this node.

#### Node Conversion

The properties in the glTF node will be converted to properties in the prefab node according to the texture relationships in the following table:

> | glTF node properties | prefab node properties |

Reviewer 2: 这个 property 跟之前属性的 property 应该是一样的，要用单数吧？

> | glTF node properties | prefab node properties |

Author: ```suggestion
| glTF Node Property | Prefab Node Property |

```

| :----------- | :----------- |
| [Hierarchy](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodechildren) | Hierarchy |
| [Displacement](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodetranslation) | Position |
| [Rotation](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#noderotation) | Rotation |
| [Scaling](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodescale) | Scaling |
| [Matrix](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodematrix) | Decompress and set the position, rotation, and scaling respectively |

> | [Mesh](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#nodemesh) | Mesh Renderer component |

Reviewer 2: Mesh Renderer 应该跟编辑器的保持一致：MeshRenderer，麻烦全文搜索检查一下，包括 Skinned Mess Renderer 和 (Skinned) Mesh Renderer

![image](https://user-images.githubusercontent.com/19702899/132791811-8180a925-6de2-4160-8de2-6155923e8d06.png)

Author: 改好了

| `TRIANGLE_STRIP` | `gfx.PrimitiveMode.TRIANGLE_STRIP` |
| `TRIANGLE_FAN`   | `gfx.PrimitiveMode.TRIANGLE_FAN`   |

> glTF vertex properties will be converted to Cocos Creator vertex properties, and the property names will be converted as shown in the following table:

Reviewer 2: ```suggestion
glTF vertex attributes will be converted to Cocos Creator vertex attributes, and the attribute names will be converted as shown in the following table:

```

> | glTF Vertex Attribute Name | Cocos Creator Vertex Attribute Name |

Reviewer 2: ```suggestion
| glTF Vertex Property Name | Cocos Creator Vertex Property Name |

```

| `JOINTS_0`                 | `gfx.AttributeName.ATTR_JOINTS`                                          |
| `WEIGHTS_0`                | `gfx.AttributeName.ATTR_WEIGHTS`                                         |

> > **NOTE**: if there are other `JOINTS`, `WEIGHTS` vertex attributes in the glTF primitive, such as `JOINTS_1`, `WEIGHTS_1`, it means that the vertices of this glTF mesh may be affected by more than 4 skeletons.

Reviewer 3: ```suggestion
> **Note**: if there are other `JOINTS`, `WEIGHTS` vertex attributes in the glTF primitive, such as `JOINTS_1`, `WEIGHTS_1`, it means that the vertices of this glTF mesh may be affected by more than 4 skeletons.

```

| `JOINTS_0`                 | `gfx.AttributeName.ATTR_JOINTS`                                          |
| `WEIGHTS_0`                | `gfx.AttributeName.ATTR_WEIGHTS`                                         |

> **Note**: if there are other `JOINTS`, `WEIGHTS` vertex attributes in the glTF primitive, such as `JOINTS_1`, `WEIGHTS_1`, it means that the vertices of this glTF mesh may be affected by more than 4 skeletons.

> For each vertex, all the weight information determined by `JOINTS_{}`, `WEIGHTS_{}` will be sorted by weight value and the four skeletons with the highest influence weight will be taken as `gfx.AttributeName.ATTR_JOINTS` and `gfx.AttributeName.ATTR_WEIGHTS`. .

Reviewer 2: ```suggestion
For each vertex, all the weight information determined by `JOINTS_{}`, `WEIGHTS_{}` will be sorted by weight value and the four skeletons with the highest influence weight will be taken as `gfx.AttributeName.ATTR_JOINTS` and `gfx.AttributeName.ATTR_WEIGHTS`.

```

| `JOINTS_0`                 | `gfx.AttributeName.ATTR_JOINTS`                                          |
| `WEIGHTS_0`                | `gfx.AttributeName.ATTR_WEIGHTS`                                         |

> **Note**: if there are other `JOINTS`, `WEIGHTS` vertex attributes in the glTF primitive, such as `JOINTS_1`, `WEIGHTS_1`, it means that the vertices of this glTF mesh may be affected by more than 4 skeletons.

For each vertex, all the weight information determined by `JOINTS_{}`, `WEIGHTS_{}` will be sorted by weight value and the four skeletons with the highest influence weight will be taken as `gfx.AttributeName.ATTR_JOINTS` and `gfx.AttributeName.ATTR_WEIGHTS`.

glTF deformation targets will be converted to Cocos Creator submesh deformation data.

### glTF Skins

After import, glTF skins will be converted to skeletal assets in Cocos Creator.

### glTF Material

After import, glTF materials will be converted to material assets in Cocos Creator.

### glTF Texture

After import, glTF texture will be converted to a texture asset in Cocos Creator.

> The glTF image referenced in the glTF texture will be converted to a reference to the corresponding converted Cocos Creator image.

Reviewer 2: 这里的 parameter 应该是 property，麻烦帮忙把中文的 “参数” 也改成 “属性”

glTF texture parameters will be converted to Cocos Creator texture parameters according to the texture in the following table:

> | glTF Texture Parameters | Cocos Creator Texture Parameters |

Reviewer 2: 这里的 parameter 应该是 property，麻烦帮忙把中文的 “参数” 也改成 “属性”

| `NEAREST_MIPMAP_LINEAR`  | `TextureBase.Filter.NEAREST` | `TextureBase.Filter.LINEAR`       |
| `NEAREST_MIPMAP_NEAREST` | `TextureBase.Filter.LINEAR`  | `TextureBase.Filter.LINEAR`       |

glTF Texture Wrap mode will be converted to Cocos Creator Texture Wrap mode according to the texture in the following table:

| glTF Texture Wrap Mode | Cocos Creator Texture Wrap Mode |
| :---------------- | :------------------------------------- |
| `CLAMP_TO_EDGE`   | `TextureBase.WrapMode.CLAMP_TO_EDGE`   |
| `REPEAT`          | `TextureBase.WrapMode.REPEAT`          |
| `MIRRORED_REPEAT` | `TextureBase.WrapMode.MIRRORED_REPEAT` |

### glTF Image

After import, the glTF image will be converted to an image asset in Cocos Creator.

> When the [URI](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#imageuri) of the glTF image is a Data URI, the image data will be fetched from the Data URI. Otherwise, the image data will be retrieved from the Data URI according to the [Cocos Creator Image Location Resolution Algorithm](./image-location-resolution.md), where `url` is the URI of the glTF image and `startDir` is the directory where the glTF file is located.

Reviewer: the image data will be resolved from the

> When the [URI](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#imageuri) of the glTF image is a Data URI, the image data will be fetched from the Data URI. Otherwise, the image data will be retrieved from the Data URI according to the [Cocos Creator Image Location Resolution Algorithm](./image-location-resolution.md), where `url` is the URI of the glTF image and `startDir` is the directory where the glTF file is located.

Author: ```suggestion
When the [URI](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#imageuri) of the glTF image is a Data URI, the image data will be fetched from the Data URI. Otherwise, the image data will be resolved from the Data URI according to the [Cocos Creator Image Location Resolution Algorithm](./image-location-resolution.md), where `url` is the URI of the glTF image and `startDir` is the directory where the glTF file is located.

```
